stablehlo/dialect/StablehloOps.cpp

```diff
@@ -4164,6 +4164,71 @@ LogicalResult RngOp::reifyReturnTypeShapes(
 // SelectOp
 //===----------------------------------------------------------------------===//
 
+namespace {
+// Utility function, used by printSelectOpType and
+// printSameOperandsAndResultType
+ParseResult assignFromFunctionType(OpAsmParser& parser, llvm::SMLoc loc,
+                                   ArrayRef<Type*> operands, Type& result,
+                                   FunctionType& fnType) {
+  assert(fnType);
+  if (fnType.getInputs().size() != operands.size()) {
+    return parser.emitError(loc)
+           << operands.size() << " operands present, but expected "
+           << fnType.getInputs().size();
+  }
+  if (fnType.getResults().size() != 1) {
+    return parser.emitError(loc, "expected single output");
+  }
+
+  // Set operand types to function input types
+  for (auto [operand, input] : llvm::zip(operands, fnType.getInputs())) {
+    *operand = input;
+  }
+  result = fnType.getResults().front();
+  return success();
+}
+}  // namespace
+
+void printSelectOpType(OpAsmPrinter& p, Operation* op, Type pred, Type onTrue,
+                       Type onFalse, Type result) {
+  // Print functional type if true/false branches don't match return type.
+  if (onTrue != result || onFalse != result) {
+    p.printFunctionalType(op);
+    return;
+  }
+
+  // Print pred type and result type
+  p << pred << ", " << result;
+}
+
+ParseResult parseSelectOpType(OpAsmParser& parser, Type& pred, Type& onTrue,
+                              Type& onFalse, Type& result) {
+  // Operand and result types are the same, use copy constructor
+  llvm::SMLoc loc = parser.getCurrentLocation();
+  SmallVector<Type> types;
+  if (parser.parseTypeList(types)) {
+    return failure();
+  }
+
+  // Error handling for invalid types
+  // Fail if not two types, or single functional type
+  if (types.size() != 2 &&
+      (types.size() != 1 || !types.front().dyn_cast<FunctionType>())) {
+    return parser.emitError(loc,
+                            "expected functional type or list of two types");
+  }
+
+  if (types.size() == 2) {
+    pred = types.front();
+    onTrue = onFalse = result = types.back();
+    return success();
+  }
+
+  FunctionType fnType = types.front().dyn_cast<FunctionType>();
+  return assignFromFunctionType(parser, loc, {&pred, &onTrue, &onFalse}, result,
+                                fnType);
+}
```

Comment:
How about restructuring this function so that each of the `types.size() == 1` and `types.size() == 2` cases is handled contiguously? E.g. right now, additional mental effort is required to realize that ending up on line 4227 means that `types.front()` must be `FunctionType`. E.g.

```
if (types.size() == 2) {
  ...
} else if (types.size() == 1 && types[0].isa<FunctionType>()) {
  auto fnType = types[0].cast<FunctionType>();
  ...
} else {
  ...
}
```

Reply:
Would be useful to have a coding standard for things like this. 

In the past I worked with a lint for "no else after return" since it's considered unnecessary, and functions could end with a return in an else statement as opposed to a top-level return.

I also tend to prefer writing validation at the beginning of a function, to act as a sort of precondition for the function. Currently there are very few functions in `StablehloOps.cpp` (I found 2) that end in the error case as opposed to the success case.

Reply:
I think I'm describing some combination of the below:

https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

Reply:
Fair enough! Let's put this on our todo list in #15, and keep the existing style in this PR.

Can you change `types[0].dyn_cast<FunctionType>()` to `types[0].cast<FunctionType>()` though? At this point, we know for sure that it's a `FunctionType`, and `cast` will help emphasize that.

Reply:
fixed